File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Core/PressureConfig.cs
using System;

namespace Iaiso.Core;

/// <summary>Thrown when a <see cref="PressureConfig"/> fails validation.</summary>
public class ConfigException : Exception
{
    public ConfigException(string message) : base(message) {}
}

/// <summary>
/// Pressure engine configuration. See <c>spec/pressure/README.md §2</c>
/// for normative ranges.
/// </summary>
public sealed class PressureConfig
{
    public double EscalationThreshold { get; }
    public double ReleaseThreshold { get; }
    public double DissipationPerStep { get; }
    public double DissipationPerSecond { get; }
    public double TokenCoefficient { get; }
    public double ToolCoefficient { get; }
    public double DepthCoefficient { get; }
    public bool PostReleaseLock { get; }

    private PressureConfig(Builder b)
    {
        EscalationThreshold = b.EscalationThreshold;
        ReleaseThreshold = b.ReleaseThreshold;
        DissipationPerStep = b.DissipationPerStep;
        DissipationPerSecond = b.DissipationPerSecond;
        TokenCoefficient = b.TokenCoefficient;
        ToolCoefficient = b.ToolCoefficient;
        DepthCoefficient = b.DepthCoefficient;
        PostReleaseLock = b.PostReleaseLock;
    }

    public static PressureConfig Defaults() => CreateBuilder().Build();
    public static Builder CreateBuilder() => new();

    public Builder ToBuilder() => new()
    {
        EscalationThreshold = EscalationThreshold,
        ReleaseThreshold = ReleaseThreshold,
        DissipationPerStep = DissipationPerStep,
        DissipationPerSecond = DissipationPerSecond,
        TokenCoefficient = TokenCoefficient,
        ToolCoefficient = ToolCoefficient,
        DepthCoefficient = DepthCoefficient,
        PostReleaseLock = PostReleaseLock,
    };

    public void Validate()
    {
        if (EscalationThreshold < 0.0 || EscalationThreshold > 1.0)
            throw new ConfigException(
                $"escalation_threshold must be in [0, 1], got {EscalationThreshold}");
        if (ReleaseThreshold < 0.0 || ReleaseThreshold > 1.0)
            throw new ConfigException(
                $"release_threshold must be in [0, 1], got {ReleaseThreshold}");
        if (ReleaseThreshold <= EscalationThreshold)
            throw new ConfigException(
                $"release_threshold must exceed escalation_threshold ({ReleaseThreshold} <= {EscalationThreshold})");
        if (TokenCoefficient < 0.0)
            throw new ConfigException(
                $"token_coefficient must be non-negative, got {TokenCoefficient}");
        if (ToolCoefficient < 0.0)
            throw new ConfigException(
                $"tool_coefficient must be non-negative, got {ToolCoefficient}");
        if (DepthCoefficient < 0.0)
            throw new ConfigException(
                $"depth_coefficient must be non-negative, got {DepthCoefficient}");
        if (DissipationPerStep < 0.0)
            throw new ConfigException(
                $"dissipation_per_step must be non-negative, got {DissipationPerStep}");
        if (DissipationPerSecond < 0.0)
            throw new ConfigException(
                $"dissipation_per_second must be non-negative, got {DissipationPerSecond}");
    }

    public sealed class Builder
    {
        public double EscalationThreshold { get; set; } = 0.85;
        public double ReleaseThreshold { get; set; } = 0.95;
        public double DissipationPerStep { get; set; } = 0.02;
        public double DissipationPerSecond { get; set; } = 0.0;
        public double TokenCoefficient { get; set; } = 0.015;
        public double ToolCoefficient { get; set; } = 0.08;
        public double DepthCoefficient { get; set; } = 0.05;
        public bool PostReleaseLock { get; set; } = true;

        public Builder WithEscalationThreshold(double v) { EscalationThreshold = v; return this; }
        public Builder WithReleaseThreshold(double v) { ReleaseThreshold = v; return this; }
        public Builder WithDissipationPerStep(double v) { DissipationPerStep = v; return this; }
        public Builder WithDissipationPerSecond(double v) { DissipationPerSecond = v; return this; }
        public Builder WithTokenCoefficient(double v) { TokenCoefficient = v; return this; }
        public Builder WithToolCoefficient(double v) { ToolCoefficient = v; return this; }
        public Builder WithDepthCoefficient(double v) { DepthCoefficient = v; return this; }
        public Builder WithPostReleaseLock(bool v) { PostReleaseLock = v; return this; }

        public PressureConfig Build() => new(this);
    }
}

File: IAIso-v5.0/core/iaiso-java/iaiso-core/src/main/java/io/iaiso/core/PressureConfig.java
package io.iaiso.core;

import java.util.Objects;

/**
 * Pressure engine configuration. See {@code spec/pressure/README.md §2}
 * for normative ranges.
 *
 * <p>The recommended pattern is to start from {@link #defaults()} and
 * use the {@linkplain Builder builder} to override fields:
 *
 * <pre>{@code
 * PressureConfig cfg = PressureConfig.builder()
 *     .escalationThreshold(0.7)
 *     .postReleaseLock(false)
 *     .build();
 * }</pre>
 */
public final class PressureConfig {
    private final double escalationThreshold;
    private final double releaseThreshold;
    private final double dissipationPerStep;
    private final double dissipationPerSecond;
    private final double tokenCoefficient;
    private final double toolCoefficient;
    private final double depthCoefficient;
    private final boolean postReleaseLock;

    private PressureConfig(Builder b) {
        this.escalationThreshold = b.escalationThreshold;
        this.releaseThreshold = b.releaseThreshold;
        this.dissipationPerStep = b.dissipationPerStep;
        this.dissipationPerSecond = b.dissipationPerSecond;
        this.tokenCoefficient = b.tokenCoefficient;
        this.toolCoefficient = b.toolCoefficient;
        this.depthCoefficient = b.depthCoefficient;
        this.postReleaseLock = b.postReleaseLock;
    }

    public double getEscalationThreshold() { return escalationThreshold; }
    public double getReleaseThreshold() { return releaseThreshold; }
    public double getDissipationPerStep() { return dissipationPerStep; }
    public double getDissipationPerSecond() { return dissipationPerSecond; }
    public double getTokenCoefficient() { return tokenCoefficient; }
    public double getToolCoefficient() { return toolCoefficient; }
    public double getDepthCoefficient() { return depthCoefficient; }
    public boolean isPostReleaseLock() { return postReleaseLock; }

    /** A {@link PressureConfig} populated with the spec-default values. */
    public static PressureConfig defaults() {
        return builder().build();
    }

    /** Validate this config. Throws {@link ConfigException} on failure. */
    public void validate() {
        if (escalationThreshold < 0.0 || escalationThreshold > 1.0) {
            throw new ConfigException(
                "escalation_threshold must be in [0, 1], got " + escalationThreshold);
        }
        if (releaseThreshold < 0.0 || releaseThreshold > 1.0) {
            throw new ConfigException(
                "release_threshold must be in [0, 1], got " + releaseThreshold);
        }
        if (releaseThreshold <= escalationThreshold) {
            throw new ConfigException(
                "release_threshold must exceed escalation_threshold (" +
                releaseThreshold + " <= " + escalationThreshold + ")");
        }
        if (tokenCoefficient < 0.0) {
            throw new ConfigException(
                "token_coefficient must be non-negative, got " + tokenCoefficient);
        }
        if (toolCoefficient < 0.0) {
            throw new ConfigException(
                "tool_coefficient must be non-negative, got " + toolCoefficient);
        }
        if (depthCoefficient < 0.0) {
            throw new ConfigException(
                "depth_coefficient must be non-negative, got " + depthCoefficient);
        }
        if (dissipationPerStep < 0.0) {
            throw new ConfigException(
                "dissipation_per_step must be non-negative, got " + dissipationPerStep);
        }
        if (dissipationPerSecond < 0.0) {
            throw new ConfigException(
                "dissipation_per_second must be non-negative, got " + dissipationPerSecond);
        }
    }

    public Builder toBuilder() {
        return new Builder()
            .escalationThreshold(escalationThreshold)
            .releaseThreshold(releaseThreshold)
            .dissipationPerStep(dissipationPerStep)
            .dissipationPerSecond(dissipationPerSecond)
            .tokenCoefficient(tokenCoefficient)
            .toolCoefficient(toolCoefficient)
            .depthCoefficient(depthCoefficient)
            .postReleaseLock(postReleaseLock);
    }

    public static Builder builder() {
        return new Builder();
    }

    public static final class Builder {
        private double escalationThreshold = 0.85;
        private double releaseThreshold = 0.95;
        private double dissipationPerStep = 0.02;
        private double dissipationPerSecond = 0.0;
        private double tokenCoefficient = 0.015;
        private double toolCoefficient = 0.08;
        private double depthCoefficient = 0.05;
        private boolean postReleaseLock = true;

        public Builder escalationThreshold(double v) { this.escalationThreshold = v; return this; }
        public Builder releaseThreshold(double v) { this.releaseThreshold = v; return this; }
        public Builder dissipationPerStep(double v) { this.dissipationPerStep = v; return this; }
        public Builder dissipationPerSecond(double v) { this.dissipationPerSecond = v; return this; }
        public Builder tokenCoefficient(double v) { this.tokenCoefficient = v; return this; }
        public Builder toolCoefficient(double v) { this.toolCoefficient = v; return this; }
        public Builder depthCoefficient(double v) { this.depthCoefficient = v; return this; }
        public Builder postReleaseLock(boolean v) { this.postReleaseLock = v; return this; }

        public PressureConfig build() {
            return new PressureConfig(this);
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof PressureConfig)) return false;
        PressureConfig other = (PressureConfig) o;
        return Double.compare(other.escalationThreshold, escalationThreshold) == 0
            && Double.compare(other.releaseThreshold, releaseThreshold) == 0
            && Double.compare(other.dissipationPerStep, dissipationPerStep) == 0
            && Double.compare(other.dissipationPerSecond, dissipationPerSecond) == 0
            && Double.compare(other.tokenCoefficient, tokenCoefficient) == 0
            && Double.compare(other.toolCoefficient, toolCoefficient) == 0
            && Double.compare(other.depthCoefficient, depthCoefficient) == 0
            && postReleaseLock == other.postReleaseLock;
    }

    @Override
    public int hashCode() {
        return Objects.hash(escalationThreshold, releaseThreshold, dissipationPerStep,
            dissipationPerSecond, tokenCoefficient, toolCoefficient, depthCoefficient,
            postReleaseLock);
    }
}
